src/providers/realtime-provider.js

```diff
@@ -162,44 +198,77 @@ export default class RealtimeProvider {
         const domainObject = subscriptionDetails.domainObject;
         const message = MESSAGES.SUBSCRIBE[domainObject.type](subscriptionDetails);
 
-        this.sendOrQueueMessage(message);
+        this.sendMessage(message);
     }
 
     sendUnsubscribeMessage(subscriptionDetails) {
         let message = MESSAGES.UNSUBSCRIBE(subscriptionDetails);
 
-        this.sendOrQueueMessage(message);
+        this.sendMessage(message);
     }
 
-    reconnect() {
-        this.subscriptionsByCall.clear();
+    #setCallFromClock(clock) {
+        const correspondingSubscription = Object.values(this.subscriptionsById).find(subscription => {
+            return subscription.domainObject.identifier.key === clock.identifier.key;
+        });
 
-        if (this.reconnectTimeout) {
-            return;
+        if (correspondingSubscription !== undefined) {
+            this.remoteClockCallNumber = correspondingSubscription.call.toString();
+        } else {
+            delete this.remoteClockCallNumber;
         }
+    }
 
-        this.reconnectTimeout = setTimeout(() => {
-            this.connect();
-            delete this.reconnectTimeout;
-        }, FALLBACK_AND_WAIT_MS[this.currentWaitIndex]);
+    #processBatchQueue(batchQueue, call) {
+        let subscriptionDetails = this.subscriptionsByCall.get(call);
+        let telemetryData = [];
 
-        if (this.currentWaitIndex < FALLBACK_AND_WAIT_MS.length - 1) {
-            this.currentWaitIndex++;
+        // possibly cancelled
+        if (!subscriptionDetails) {
+            return;
         }
-    }
 
-    sendOrQueueMessage(request) {
-        if (this.connected) {
-            try {
-                this.sendMessage(request);
-            } catch (error) {
-                this.connected = false;
-                this.requests.push(request);
-                console.error("🚨 Error while attempting to send to websocket, closing websocket", error);
-                this.socket.close();
-            }
-        } else {
-            this.requests.push(request);
+        batchQueue.forEach((rawMessage) => {
+            const message = JSON.parse(rawMessage);
+            const values = message.data.values || [];
+            const parentName = subscriptionDetails.domainObject.name;
+
+            values.forEach(parameter => {
+                let datum = {
+                    timestamp: parameter[METADATA_TIME_KEY]
+                };
+                const value = getValue(parameter, parentName);
+
+                // TODO: optimize this. I think we only care if the last value is stale.
+                if (this.observingStaleness[subscriptionDetails.name] !== undefined) {
```

Comment:
Agreed it's not ideal. I'm trying to avoid the overhead of making function calls in this high-frequency code, and I actually removed some existing function calls here.

This highlights the fact that I've left a todo here. It's an optimization though, and I think I'll file a followup because changing it now will significantly increase the regression testing needed on this PR.

Reply:
I will quote Mr. Henry on this one and say "Remove the comment and file an issue" 😁

Reply:
Whoever this Mr Henry is, he's a wise man.

Reply:
Created https://github.com/akhenry/openmct-yamcs/issues/416